torch/distributed/_composable/fsdp/_fsdp_collectives.py

```diff
@@ -179,6 +177,58 @@ def foreach_all_gather(
         )
 
 
+@torch.no_grad()
+def _get_param_all_gather_inputs(
+    fsdp_params: List[FSDPParam],
+) -> List[List[torch.Tensor]]:
+    if ca.compiled_autograd_enabled:
+        return [fsdp_param.all_gather_inputs for fsdp_param in fsdp_params]
+
+    # Intentionally try to run a fast-path that bypasses abstractions for the
+    # common FSDP case of bf16/fp32 mixed precision in order to use foreach
+    # copy for lower CPU overhead and more efficient copying in eager
+    def use_foreach_copy(fsdp_param: FSDPParam) -> bool:
+        return (
+            fsdp_param.param_dtype is not None
+            and not fsdp_param.offload_to_cpu
```

Comment:
We could make this so that even if we do CPU offloading, then we can still do the H2D first and then this foreach copy.

I guess we can do this in a follow-up if we really need it in the future.